community/kernel/src/main/java/org/neo4j/kernel/impl/core/DelegatingTokenHolder.java

```diff
@@ -112,6 +121,76 @@ private synchronized int createToken( String name ) throws KernelException
         return id;
     }
 
+    @Override
+    public void getOrCreateIds( String[] names, int[] ids )
+    {
+        if ( names.length != ids.length )
+        {
+            throw new IllegalArgumentException( "Name and id arrays must have the same length." );
+        }
+        // Assume all tokens exist and try to resolve them. Break out on the first missing token.
+        boolean hasUnresolvedTokens = resolveIds( names, ids, ALWAYS_TRUE_INT );
+
+        if ( hasUnresolvedTokens )
+        {
+            createMissingTokens( names, ids );
+        }
+    }
+
+    private boolean resolveIds( String[] names, int[] ids, IntPredicate unresolvedIndexCheck )
+    {
+        boolean foundUnresolvable = false;
+        for ( int i = 0; i < ids.length; i++ )
+        {
+            Integer id = tokenCache.getId( names[i] );
+            if ( id != null )
+            {
+                ids[i] = id;
+            }
+            else
+            {
+                foundUnresolvable = true;
+                if ( unresolvedIndexCheck.test( i ) )
+                {
+                    // If the check returns `true`, it's a signal that we should stop early.
+                    break;
+                }
+            }
+        }
+        return foundUnresolvable;
+    }
+
+    private synchronized void createMissingTokens( String[] names, int[] ids )
+    {
+        // We redo the resolving under the lock, to make sure that these ids are really missing, and won't be
+        // created concurrently with us.
+        MutableIntSet unresolvedIndexes = new IntHashSet();
+        resolveIds( names, ids, i -> !unresolvedIndexes.add( i ) );
```

Comment:
A name duplicate would make resolveIds stop early, which would result in not all names being resolved/created. Problem?

Reply:
And if we pass it through… might duplicate unresolved names be created twice? I'm conflicted about whether to handle it, or document that there must be no duplicates. Any opinions from @systay or @fickludd?

Reply:
I don't know how Cypher would use this batching atm. In the situation described in the github issue we would be guaranteed to have to duplicates because the properties are provided in a map. 

It feels to me like handling duplicates nicely will be a small cost compared to the issue we are solving, so maybe just handle it?